### ISSUE TITLE
Experiment with map_overlap and cupy arrays

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#4803]
It would be useful to try the [da.map_overlap](http://docs.dask.org/en/latest/array-overlap.html) function with CuPy arrays on large 2d datasets.

A trivial example might look something like the following (untested):

```python
import dask.array as da
import cupy

rs = da.random.RandomState(RandomState=cupy.random.RandomState)  # swap cupy->numpy here for comparison
x = rs.random(500000, 500000), chunks=(10000, 10000))

x.map_overlap(lambda x: x, depth=1)
x.sum().compute()  # trigger computation, but don't ask for the entire array as a result
```

My guess is that we'll be badly bound by communication.  I would verify this probably by running this computation under the [dask distributed scheduler](http://docs.dask.org/en/latest/scheduling.html#dask-distributed-local) but started with [dask-cuda's](https://github.com/rapidsai/dask-cuda) LocalCUDACluster, and then by watching the [dashboard](https://docs.dask.org/en/latest/diagnostics-distributed.html#dashboard).  

My hope is that once the UCX work finishes that this cost goes down considerably.  It will be interesting to see by how much.

Additionally, we might try using numba.cuda.jit to build some simple nearest-neighbor kernel function and applying that with `map_overlap` over the array.  [This notebook](https://gist.github.com/mrocklin/9272bf84a8faffdbbe2cd44b4bc4ce3c) from [this blogpost](https://blog.dask.org/2019/04/09/numba-stencil) might be an interesting starting point here (but there are probably more interesting operations).

cc @madsbk

**Comment by mrocklin** [comment]
Also cc @jakirkham and @pentschev for visibility

**Comment by jakirkham** [comment]
Thanks for raising this. This would be good to look into. Imagine array creation becomes pretty important here (as Peter has discovered in other parts of the codebase).

We may also want to take a closer look at `da.pad`. It handles a larger variety of edge conditions and also does some array creation in the process.

**Comment by mrocklin** [comment]
Nice points.  Thanks @jakirkham  !

**Comment by pentschev** [comment]
I've done some testing before. Without UCX we're in many cases better off using a CPU-only implementation on a single process. The major problem is having different processes, where in practice we can go no faster than ~1GB/s with IPC via TCP.

Only for some context, CUDA performance is often bound by PCIe communication, where theoretical bandwidth is 15.75GB/s, with 10-12 GB/s being achievable in practice on my own experience. So we're a whole order of magnitude down when doing any IPC via TCP.

**Comment by jakirkham** [comment]
Just to add an observation here, currently the logic in `map_overlap` (particularly around constant edge values) runs into some issues as constant edge values are effectively turned into NumPy arrays, which cannot be concatenated with CuPy arrays. So we will probably want to rework this logic.

```python
r = a_gpu.map_overlap(lambda x: x, depth=1, boundary=0)
r.compute()
```

<details>

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-070ae56efcaf> in <module>
----> 1 r.compute()

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/base.py in compute(self, **kwargs)
    154         dask.base.compute
    155         """
--> 156         (result,) = compute(self, traverse=False, **kwargs)
    157         return result
    158 

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/base.py in compute(*args, **kwargs)
    396     keys = [x.__dask_keys__() for x in collections]
    397     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 398     results = schedule(dsk, keys, **kwargs)
    399     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    400 

~/miniconda/envs/img_exp/lib/python3.7/site-packages/distributed/client.py in get(self, dsk, keys, restrictions, loose_restrictions, resources, sync, asynchronous, direct, retries, priority, fifo_timeout, actors, **kwargs)
   2566                     should_rejoin = False
   2567             try:
-> 2568                 results = self.gather(packed, asynchronous=asynchronous, direct=direct)
   2569             finally:
   2570                 for f in futures.values():

~/miniconda/envs/img_exp/lib/python3.7/site-packages/distributed/client.py in gather(self, futures, errors, maxsize, direct, asynchronous)
   1820                 direct=direct,
   1821                 local_worker=local_worker,
-> 1822                 asynchronous=asynchronous,
   1823             )
   1824 

~/miniconda/envs/img_exp/lib/python3.7/site-packages/distributed/client.py in sync(self, func, *args, **kwargs)
    751             return future
    752         else:
--> 753             return sync(self.loop, func, *args, **kwargs)
    754 
    755     def __repr__(self):

~/miniconda/envs/img_exp/lib/python3.7/site-packages/distributed/utils.py in sync(loop, func, *args, **kwargs)
    329             e.wait(10)
    330     if error[0]:
--> 331         six.reraise(*error[0])
    332     else:
    333         return result[0]

~/miniconda/envs/img_exp/lib/python3.7/site-packages/six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

~/miniconda/envs/img_exp/lib/python3.7/site-packages/distributed/utils.py in f()
    314             if timeout is not None:
    315                 future = gen.with_timeout(timedelta(seconds=timeout), future)
--> 316             result[0] = yield future
    317         except Exception as exc:
    318             error[0] = sys.exc_info()

~/miniconda/envs/img_exp/lib/python3.7/site-packages/tornado/gen.py in run(self)
   1131 
   1132                     try:
-> 1133                         value = future.result()
   1134                     except Exception:
   1135                         self.had_exception = True

~/miniconda/envs/img_exp/lib/python3.7/site-packages/tornado/gen.py in run(self)
   1139                     if exc_info is not None:
   1140                         try:
-> 1141                             yielded = self.gen.throw(*exc_info)
   1142                         finally:
   1143                             # Break up a reference to itself

~/miniconda/envs/img_exp/lib/python3.7/site-packages/distributed/client.py in _gather(self, futures, errors, direct, local_worker)
   1651                             six.reraise(CancelledError, CancelledError(key), None)
   1652                         else:
-> 1653                             six.reraise(type(exception), exception, traceback)
   1654                     if errors == "skip":
   1655                         bad_keys.add(key)

~/miniconda/envs/img_exp/lib/python3.7/site-packages/six.py in reraise(tp, value, tb)
    690                 value = tp()
    691             if value.__traceback__ is not tb:
--> 692                 raise value.with_traceback(tb)
    693             raise value
    694         finally:

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in concatenate3()
   3605     if concatenate_lookup.dispatch(type(advanced)) is not np.concatenate:
   3606         x = unpack_singleton(arrays)
-> 3607         return _concatenate2(arrays, axes=list(range(x.ndim)))
   3608 
   3609     ndim = ndimlist(arrays)

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in _concatenate2()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in <listcomp>()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in _concatenate2()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in <listcomp>()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in _concatenate2()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in <listcomp>()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in _concatenate2()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in <listcomp>()
    224         return arrays
    225     if len(axes) > 1:
--> 226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
    228     return concatenate(arrays, axis=axes[0])

~/miniconda/envs/img_exp/lib/python3.7/site-packages/dask/array/core.py in _concatenate2()
    226         arrays = [_concatenate2(a, axes=axes[1:]) for a in arrays]
    227     concatenate = concatenate_lookup.dispatch(type(max(arrays, key=lambda x: x.__array_priority__)))
--> 228     return concatenate(arrays, axis=axes[0])
    229 
    230 

~/miniconda/envs/img_exp/lib/python3.7/site-packages/cupy/manipulation/join.py in concatenate()
     47 
     48     """
---> 49     return core.concatenate_method(tup, axis)
     50 
     51 

cupy/core/_routines_manipulation.pyx in cupy.core._routines_manipulation.concatenate_method()

cupy/core/_routines_manipulation.pyx in cupy.core._routines_manipulation.concatenate_method()

TypeError: Only cupy arrays can be concatenated
```

</details>

**Comment by pentschev** [comment]
Ok, I played around a bit with `map_overlap` and UCX. For the sample in https://gist.github.com/pentschev/d0aa9cd20d543c3441e530e8067eb765#file-ucx_map_overlap_smooth-py, I see the compute time at around 19-21 seconds when using UCX, versus 33-35 seconds when simply using `LocalCUDACluster`.

One thing I noticed is that each time I run the code, one of the GPUs consumes a much larger amount of memory than others (at a ratio around 3:1), thus, increasing the size much beyond `(20000, 20000)` ends up OOM. The same happens no matter what function is passed to `map_overlap` (including a simple `lambda x: x`). I'm not sure why that happens, could someone explain this?

**Comment by mrocklin** [comment]
```python
            result = await client.compute(y)
```

When you call `compute` you're asking for a non-dask result locally in your process.  Instead, you might persist the dataframe and then await it.  This will keep things in distributed memory

```python
result = await y.persist()
```

Additionally, you might consider using smaller chunks than what is here:

```python
            x = rs.random((20000, 20000), chunks=(10000, 10000)).persist()
```

I wonder if this would happen with smaller chunks

**Comment by mrocklin** [comment]
> I see the compute time at around 19-21 seconds when using UCX, versus 33-35 seconds when simply using LocalCUDACluster.

Along with speedup numbers, it might also be good to get a sense for what the current bottlenecks are.  Are we bound by communication?  If so, what do those numbers look like?

**Comment by pentschev** [comment]
> ```python
>             result = await client.compute(y)
> ```
> 
> When you call `compute` you're asking for a non-dask result locally in your process. Instead, you might persist the dataframe and then await it. This will keep things in distributed memory
> 
> ```python
> result = await y.persist()
> ```

Right, that one was obvious, I should have seen that. Thanks for pointing it out!

> Additionally, you might consider using smaller chunks than what is here:
> 
> ```python
>             x = rs.random((20000, 20000), chunks=(10000, 10000)).persist()
> ```
> 
> I wonder if this would happen with smaller chunks

I tried also smaller chunks, there was a change but wasn't too significant (the ratio got around 3.5:1 for `(5000, 5000)`).

> Along with speedup numbers, it might also be good to get a sense for what the current bottlenecks are. Are we bound by communication? If so, what do those numbers look like?

The current bottleneck is almost certainly communication (or, potentially, tasks related to that, such as serialization). There are two main reasons that lead me to this conclusion:

1. By just moving from `LocalCUDACluster` to `DGX` we slash compute time by around 40%;
1. This is harder to quantify and present numerically, but watching `nvidia-smi` during execution shows that GPU utilization is critically low (never seems to go above 10-15%).

**Comment by pentschev** [comment]
After replacing `await client.compute(y)` by `await y.persist()` I was able to go to array size of `(80000, 80000)`. Results below are still consistent with my previous comment (likely bound by communication).

UCX (`chunks=(10000, 10000)`): 6-7 seconds
UCX (`chunks=(5000, 5000)`): 21-24 seconds
TCP (`chunks=(10000, 10000)`): 17-23 seconds
TCP (`chunks=(5000, 5000)`): 53-58 seconds

Note: probably obvious, but TCP uses `LocalCUDACluster`, whereas UCX uses `DGX`.

**Comment by jakirkham** [comment]
I'd be curious to know how much of that time is spent by Dask doing something vs. data transmission.

**Comment by pentschev** [comment]
I've also updated [the gist](https://gist.github.com/pentschev/d0aa9cd20d543c3441e530e8067eb765#file-ucx_map_overlap_smooth-py) now to reflect the setup I used to obtain the results from my previous comment.

**Comment by pentschev** [comment]
> I'd be curious to know how much of that time is spent by Dask doing something vs. data transmission.

Given that GPU utilization doesn't go over 15%, I'd say probably 10-20% of the time is actually on compute. Right now, I'm not exactly sure how to get accurate measurements of time spent in communication.

**Comment by jakirkham** [comment]
It might be worth comparing a `map_blocks` case to the `map_overlap` case. That would at least give us and idea how much the overlap is costing in different cases.

**Comment by mrocklin** [comment]
You might consider using get_task_stream to gather coarse grained data
about compute and communication tasks.
https://docs.dask.org/en/latest/futures.html#distributed.Client.get_task_stream

This doesn't give much low level detail though

**Comment by mrocklin** [comment]
Also, are you using all eight GPUs, or just 0-3 or 4-7 ?

**Comment by pentschev** [comment]
> Also, are you using all eight GPUs, or just 0-3 or 4-7 ?

When you use `DGX`, it automatically uses 0-3. Nevertheless, I've set to use only 4 of them (4-7 in my case, since others were in use) in all cases to get a fair comparison.

**Comment by pentschev** [comment]
Quick update on using `map_blocks`:

UCX (chunks=(10000, 10000)): 1.9 seconds (vs 6-7 seconds with `map_overlap`)
UCX (chunks=(5000, 5000)): 6.5 seconds (vs 21-24 seconds with `map_overlap`)
TCP (chunks=(10000, 10000)): 1.8 seconds (vs 17-23 seconds with `map_overlap`)
TCP (chunks=(5000, 5000)): 6.2 seconds (vs 53-58 seconds with `map_overlap`)

**Comment by pentschev** [comment]
I also saved the results of task streams for `map_overlap`, they can be found in https://gist.github.com/pentschev/b0977b9d4eae3bea915ee3de617cf049 (sorry, they're html files which need to be downloaded and opened locally, I don't know of a better way to share this). Spoiler alert: data transfer with UCX is considerably faster than with TCP.

I didn't share the task streams for `map_blocks` because they're really uninteresting, there seems to be no communication happening at all, so they're mostly equal for both UCX and TCP.

**Comment by jakirkham** [comment]
Maybe it's worth even picking a simpler computation just to further isolate the communication between `map_blocks` and `map_overlap` cases (for example `lambda x: x`).

**Comment by pentschev** [comment]
Results with `lambda x: x`

UCX (`chunks=(10000, 10000)`): 3.6 seconds (vs 6-7 seconds with Numba smooth kernel)
UCX (`chunks=(5000, 5000)`): 11.0 seconds (vs 21-24 seconds with Numba smooth kernel)
TCP (`chunks=(10000, 10000)`): 23.9 seconds (vs 17-23 seconds with Numba smooth kernel)
TCP (`chunks=(5000, 5000)`): 16.3 seconds (vs 53-58 seconds with Numba smooth kernel)

Task stream gist: https://gist.github.com/pentschev/24feb780a4ae857b31f21827386527b6

It's funny to look at the task streams for this case, they're almost entirely communication for both TCP and UCX.

**Comment by jakirkham** [comment]
It's interesting to see the `map_blocks` time went up significantly for the TCP case with that change. Do we know why that is?

**Comment by mrocklin** [comment]
The easiest way I've found to do this is to push the HTML files to a gh-pages branch and then go to username.github.io/repo-name/path-to-file.html

http://mrocklin.github.io/raw-host/map-overlap/map_overlap_10k_tcp.html

http://mrocklin.github.io/raw-host/map-overlap/map_overlap_10k_ucx.html

https://github.com/mrocklin/raw-host/commit/72f0876b88c2f7d4e4bc0b5e845811a28fc220cc

**Comment by mrocklin** [comment]
I'm seeing 1s transfers, which seems quite slow to me given that we're moving small data around.

@pentschev one thing that you might try is to have the workers all in the same process and then profile things with cProfile.  This would be a little bit unrealistic, because all of the workers would be using the same GPU, but whatever is slowing down the transfers in the real case might also slow down things in the single-GPU case.  To do this, I would use one of the following:

```python
async with LocalCluster(processes=False, n_workers=4, threads_per_worker=1, asynchronous=True) as cluster:
    ...
```

Or this

```python
async with Scheduler() as s:
    async with Worker(s.address, nthreads=1), Worker(s.address, nthreads=1):
        ...
```

Then I would run the Python script with the cProfile module

```
python -m cProfile -o my-profile.out myscript.py
```

And then visualize the results with snakeviz

```
pip install snakeviz
snakeviz my-profile.out
```
       
There will probably be lots of information about Tornado and event loops, which is probably hard to parse, but somewhere in there might also be a nice bundle of information about UCX transfers.

Alternatively, you might also consider navigating to one of the worker dashboards (these should be accessible from the info pages for each worker) and then going to the `/profile-server` route, which should give you a flame graph profile plot of the main administrative thread of the worker.

I'm happy to help out with any of this tomorrow.  It probably doesn't make sense for you to spend much time figuring out these things if you don't already know them.

**Comment by pentschev** [comment]
> It's interesting to see the map_blocks time went up significantly for the TCP case with that change. Do we know why that is?

Can you point exactly which result you're referring to? If it was about my edit in https://github.com/dask/dask/issues/4803#issuecomment-520594710, times doubled as we halved the number of GPUs, which makes sense to me, given that that operation is bound by compute rather than communication.

**Comment by pentschev** [comment]
@mrockling I've used snakeviz before, let me try and see if I find something interesting, if not, we can later today discuss what else we can do to profile it better.

**Comment by mrocklin** [comment]
Note that in my examples above I failed to add `protocol=` and `interface=` keywords to the scheduler and workers.

**Comment by mrocklin** [comment]
@pentschev is going on holiday for a couple of weeks.

@madsbk do you have some time to take a look at this?  It would be good to experiment a bit with sizes here, and see what is taking time.

**Comment by martindurant** [comment]
ping on this issue - seems to be a valuable one to get right

**Comment by jakirkham** [comment]
I could be wrong, but I think this works today, which wasn't originally the case. Saying this because PR ( https://github.com/dask/dask-image/pull/151 ) recently added GPU support for dask-image filters, which use CuPy with `map_overlap`.

There's a separate question about performance. Particularly as it relates to communication overhead. It's probably worth just running things again to see where things stand (so much has changed in the last year! 😄).

cc @GenevieveBuckley (in case you know of current issues here that should be looked at)

**Comment by GenevieveBuckley** [comment]
I don't know of additional issues, but the biggest blocker for us is array creation. Maybe the `*_like` functions could be extended to support an array type as well as a dtype? https://github.com/dask/dask-image/issues/133#issuecomment-685404678

I haven't done any performance benchmarking, that would be very valuable to look at again (@jakirkham is right, a lot might have changed in a year).

**Comment by pentschev** [comment]
This issue was mostly around improving communication performance. @jakirkham is right, this already works for CuPy arrays, and communication can be improved with [UCX-Py](https://ucx-py.readthedocs.io/en/latest/) provided the running system has the specialized hardware for it. I'm pretty sure we could eventually get updated numbers to those already posted above and I'm confident they are now much better than they were in 2019. Note that we have a few [benchmarks in dask-cuda](https://github.com/rapidsai/dask-cuda/tree/branch-0.16/dask_cuda/benchmarks) where people are able to test TCP vs UCX communication, which shows what we once were looking with code from past comments.

It seems though that this issue can be misleading to people, where a possible interpretation is that this didn't work at all rather than the desire to improve performance, so perhaps the best is to close it?

**Comment by jakirkham** [comment]
Have gone ahead and put together a simple benchmark script in PR ( https://github.com/rapidsai/dask-cuda/pull/399 ). This should give us a way to track performance and measure improvements. Perhaps we can close this once that is in? Further issues could follow up on more specific improvements as needed.
